<h1><center>HIV DATASET ANALYSIS</center></h1>

## Author : Bhupender Singh

## MKR Task - 1

### 📌 Problem Statement  - To Clean, Analyze and Visualize The Given HIV Dataset


### Let's Start With Importing The Necessary Modules For This Data Cleaning Task

In [1]:
import pandas as pd
import numpy as np
print('Modules Succesfully Imported')

Modules Succesfully Imported


<h1><center>Sheet-1</center></h1>

### Importing The First Dataset

In [2]:
df1 = pd.read_excel('HIV-0000000026,SDGHIV (1).xlsx')

In [3]:
df1.head()

Unnamed: 0 Unnamed: 1     Number of new HIV infections  \
0  WHO region       Year                                    
1      Global       2019  1 700 000 [1 200 000–2 200 000]   
2      Global       2018  1 700 000 [1 200 000–2 300 000]   
3      Global       2017  1 800 000 [1 300 000–2 400 000]   
4      Global       2016  1 800 000 [1 300 000–2 400 000]   

  New HIV infections (per 1000 uninfected population)  \
0                                         Both sexes    
1                                    0.22 [0.16–0.3]    
2                                   0.23 [0.17–0.31]    
3                                   0.24 [0.18–0.32]    
4                                   0.25 [0.18–0.34]    

  New HIV infections (per 1000 uninfected population).1  \
0                                               Male      
1                                   0.23 [0.17–0.32]      
2                                   0.24 [0.17–0.33]      
3                                   0.24 [0.18–0.34]      
4                                   0.25 [0.18–0.35]      

  New HIV infections (per 1000 uninfected population).2  
0                                             Female     
1                                   0.21 [0.16-0.29]     
2                                    0.22 [0.16-0.3]     
3                                   0.23 [0.17-0.32]     
4                                   0.24 [0.18-0.33]

In [4]:
df1.shape

(211, 6)

**Dataset Contains 211 Rows and 6 Columns**

**Now Removing The Brackets and Values Inside Them To Get The Exact Value**

In [5]:
df1['Number of new HIV infections'] = df1['Number of new HIV infections'].str.split('[').str[0]

df1['New HIV infections (per 1000 uninfected population)'] = df1['New HIV infections (per 1000 uninfected population)'].str.split('[').str[0]

df1['New HIV infections (per 1000 uninfected population).1'] = df1['New HIV infections (per 1000 uninfected population).1'].str.split('[').str[0]

df1['New HIV infections (per 1000 uninfected population).2'] = df1['New HIV infections (per 1000 uninfected population).2'].str.split('[').str[0]

df1.head()

Unnamed: 0 Unnamed: 1 Number of new HIV infections  \
0  WHO region       Year                                
1      Global       2019                   1 700 000    
2      Global       2018                   1 700 000    
3      Global       2017                   1 800 000    
4      Global       2016                   1 800 000    

  New HIV infections (per 1000 uninfected population)  \
0                                         Both sexes    
1                                              0.22     
2                                              0.23     
3                                              0.24     
4                                              0.25     

  New HIV infections (per 1000 uninfected population).1  \
0                                               Male      
1                                              0.23       
2                                              0.24       
3                                              0.24       
4                                              0.25       

  New HIV infections (per 1000 uninfected population).2  
0                                             Female     
1                                              0.21      
2                                              0.22      
3                                              0.23      
4                                              0.24

**We can see that at index 0 some Part of column names are present So. Let's Fix this**

In [6]:
df1.drop(df1.index[0], inplace = True)

**Renaming The Columns After Fixing The Index 0**

In [7]:
df1 = df1.rename(columns={'Unnamed: 0': 'WHO region', 'Unnamed: 1': 'Year',
                         'New HIV infections (per 1000 uninfected population)' : 'New HIV infections (per 1000 uninfected population) (Both Sexes)',
                         'New HIV infections (per 1000 uninfected population).1' : 'New HIV infections (per 1000 uninfected population) (Male)',
                         'New HIV infections (per 1000 uninfected population).2' : 'New HIV infections (per 1000 uninfected population) (Female)'})

**Removing the whitespace between the numbers**

In [8]:
for col in df1.columns:
    if col!='Year':
        df1[col]=df1[col].str.replace(' ','')

In [9]:
df1.head()

WHO region  Year Number of new HIV infections  \
1     Global  2019                      1700000   
2     Global  2018                      1700000   
3     Global  2017                      1800000   
4     Global  2016                      1800000   
5     Global  2015                      1900000   

  New HIV infections (per 1000 uninfected population) (Both Sexes)  \
1                                               0.22                 
2                                               0.23                 
3                                               0.24                 
4                                               0.25                 
5                                               0.26                 

  New HIV infections (per 1000 uninfected population) (Male)  \
1                                               0.23           
2                                               0.24           
3                                               0.24           
4                                               0.25           
5                                               0.26           

  New HIV infections (per 1000 uninfected population) (Female)  
1                                               0.21            
2                                               0.22            
3                                               0.23            
4                                               0.24            
5                                               0.26

In [10]:
df1.shape

(210, 6)

**First Dataset is cleaned and the Final Cleaned Dataset Contains 210 rows and 6 columns**

<h1><center>Sheet-2</center></h1>

### Importing the second dataset

In [11]:
df2 = pd.read_excel('HIV-0000000006.xlsx')

In [12]:
df2.head()

WHO region  Year Number of deaths due to HIV/AIDS
0     Global  2019        690 000 [500 000–970 000]
1     Global  2018      730 000 [530 000–1 000 000]
2     Global  2017      760 000 [550 000–1 100 000]
3     Global  2016      800 000 [580 000–1 100 000]
4     Global  2015      830 000 [610 000–1 200 000]

**Removing the brackets and values inside them to get the exact value**

In [13]:
df2['Number of deaths due to HIV/AIDS'] = df2['Number of deaths due to HIV/AIDS'].str.split('[').str[0]

df2.head()

WHO region  Year Number of deaths due to HIV/AIDS
0     Global  2019                         690 000 
1     Global  2018                         730 000 
2     Global  2017                         760 000 
3     Global  2016                         800 000 
4     Global  2015                         830 000

In [14]:
df2.shape

(210, 3)

**Removing the whitespace between the numbers**

In [15]:
for col in df2.columns:
    if col!='Year':
        df2[col]=df2[col].str.replace(' ','')

In [16]:
df2.head()

WHO region  Year Number of deaths due to HIV/AIDS
0     Global  2019                           690000
1     Global  2018                           730000
2     Global  2017                           760000
3     Global  2016                           800000
4     Global  2015                           830000

In [17]:
df2.shape

(210, 3)

**The second dataset is cleaned and the cleaned dataset contains 210 rows and 3 columns**

<h1><center>Sheet-3</center><h1>

**Importing the third dataset**

In [18]:
df3 = pd.read_excel('HIV-0000000001 (2).xlsx')

In [19]:
df3.head()

REGION  YEAR PEOPLE HAVING HIV
0  Global  2019       38 000 000 
1  Global  2018        37 300 000
2  Global  2017        36 500 000
3  Global  2016        35 700 000
4  Global  2015       34 900 000

**Renaming The Columns as They are Similar to WHO region & Year Column In Sheet 1 and This Also Helps In Joining/Merging The Sheets**

In [20]:
df3 = df3.rename(columns={'REGION': 'WHO region', 'YEAR': 'Year'})

In [21]:
df3.head()

WHO region  Year PEOPLE HAVING HIV
0     Global  2019       38 000 000 
1     Global  2018        37 300 000
2     Global  2017        36 500 000
3     Global  2016        35 700 000
4     Global  2015       34 900 000

In [22]:
df3.shape

(210, 3)

**Removing the Whitespace between the numbers**

In [23]:
for col in df3.columns:
    if col!='Year':
        df3[col]=df3[col].str.replace(' ','')

In [24]:
df3.shape

(210, 3)

**The Third dataset is cleaned and The cleaned Dataset contains 210 rows and columns**

<h1><center>Merging The Sheets</center></h1>

### Merging Sheets 1 & 2

In [25]:
df_merge1 = pd.merge(df1,df2, on=['WHO region', 'Year'], how='inner')

In [26]:
df_merge1.head()

WHO region  Year Number of new HIV infections  \
0     Global  2019                      1700000   
1     Global  2018                      1700000   
2     Global  2017                      1800000   
3     Global  2016                      1800000   
4     Global  2015                      1900000   

  New HIV infections (per 1000 uninfected population) (Both Sexes)  \
0                                               0.22                 
1                                               0.23                 
2                                               0.24                 
3                                               0.25                 
4                                               0.26                 

  New HIV infections (per 1000 uninfected population) (Male)  \
0                                               0.23           
1                                               0.24           
2                                               0.24           
3                                               0.25           
4                                               0.26           

  New HIV infections (per 1000 uninfected population) (Female)  \
0                                               0.21             
1                                               0.22             
2                                               0.23             
3                                               0.24             
4                                               0.26             

  Number of deaths due to HIV/AIDS  
0                           690000  
1                           730000  
2                           760000  
3                           800000  
4                           830000

In [27]:
df_merge1.shape

(210, 7)

### Merging All Three Together 

In [28]:
df_merge2 = pd.merge( df_merge1, df3, on=['WHO region','Year'], how='inner')

In [29]:
df_merge2.head()

WHO region  Year Number of new HIV infections  \
0     Global  2019                      1700000   
1     Global  2018                      1700000   
2     Global  2017                      1800000   
3     Global  2016                      1800000   
4     Global  2015                      1900000   

  New HIV infections (per 1000 uninfected population) (Both Sexes)  \
0                                               0.22                 
1                                               0.23                 
2                                               0.24                 
3                                               0.25                 
4                                               0.26                 

  New HIV infections (per 1000 uninfected population) (Male)  \
0                                               0.23           
1                                               0.24           
2                                               0.24           
3                                               0.25           
4                                               0.26           

  New HIV infections (per 1000 uninfected population) (Female)  \
0                                               0.21             
1                                               0.22             
2                                               0.23             
3                                               0.24             
4                                               0.26             

  Number of deaths due to HIV/AIDS PEOPLE HAVING HIV  
0                           690000          38000000  
1                           730000          37300000  
2                           760000          36500000  
3                           800000          35700000  
4                           830000          34900000

In [30]:
df_merge2.shape

(210, 8)

In [31]:
df_merge2.tail()

WHO region  Year Number of new HIV infections  \
205  WesternPacific  1994                       140000   
206  WesternPacific  1993                        50000   
207  WesternPacific  1992                        40000   
208  WesternPacific  1991                        30000   
209  WesternPacific  1990                        23000   

    New HIV infections (per 1000 uninfected population) (Both Sexes)  \
205                                               0.09                 
206                                               0.03                 
207                                               0.03                 
208                                               0.02                 
209                                               0.02                 

    New HIV infections (per 1000 uninfected population) (Male)  \
205                                               0.13           
206                                               0.05           
207                                               0.04           
208                                               0.03           
209                                               0.02           

    New HIV infections (per 1000 uninfected population) (Female)  \
205                                               0.05             
206                                               0.02             
207                                               0.01             
208                                           &lt;0.01             
209                                           &lt;0.01             

    Number of deaths due to HIV/AIDS PEOPLE HAVING HIV  
205                             7700            320000  
206                             5500            200000  
207                             4000            150000  
208                             2900            120000  
209                             2100             90000

In [32]:
df_merge2.columns

Index(['WHO region', 'Year', 'Number of new HIV infections',
       'New HIV infections (per 1000 uninfected population) (Both Sexes)',
       'New HIV infections (per 1000 uninfected population) (Male)',
       'New HIV infections (per 1000 uninfected population) (Female)',
       'Number of deaths due to HIV/AIDS', 'PEOPLE HAVING HIV'],
      dtype='object')

**Changing some values eg. & lt;0.01 to the closest value**

In [33]:
df_merge2['New HIV infections (per 1000 uninfected population) (Both Sexes)'].unique()

array(['0.22', '0.23', '0.24', '0.25', '0.26', '0.27', '0.28', '0.29',
       '0.3', '0.32', '0.33', '0.34', '0.35', '0.36', '0.37', '0.38',
       '0.4', '0.42', '0.43', '0.45', '0.47', '0.48', '0.49', '0.41',
       '0.94', '1', '1.09', '1.17', '1.26', '1.36', '1.46', '1.57', '1.7',
       '1.83', '1.94', '2.06', '2.18', '2.3', '2.41', '2.53', '2.66',
       '2.8', '2.94', '3.09', '3.22', '3.34', '3.44', '3.51', '3.57',
       '3.59', '3.54', '3.42', '3.21', '2.92', '0.17', '0.18', '0.19',
       '0.2', '0.21', '0.08', '0.09', '0.1', '0.11', '0.12', '0.13',
       '0.14', '0.15', '0.16', '0.07', '0.06', '0.05', '0.04', '0.03',
       '0.02', '0.01', '&lt;0.01'], dtype=object)

In [34]:
df_merge2['New HIV infections (per 1000 uninfected population) (Female)'].unique()

array(['0.21', '0.22', '0.23', '0.24', '0.26', '0.27', '0.28', '0.29',
       '0.31', '0.32', '0.34', '0.35', '0.36', '0.37', '0.39', '0.4',
       '0.41', '0.43', '0.45', '0.47', '0.48', '0.49', '0.5', '0.44',
       '1.11', '1.18', '1.29', '1.38', '1.48', '1.6', '1.71', '1.85',
       '1.99', '2.14', '2.26', '2.39', '2.53', '2.66', '2.79', '2.92',
       '3.07', '3.23', '3.4', '3.56', '3.71', '3.84', '3.95', '4.02',
       '4.07', '4.08', '3.99', '3.83', '3.55', '3.19', '0.09', '0.1',
       '0.11', '0.12', '0.13', '0.14', '0.15', '0.06', '0.07', '0.08',
       '0.16', '0.19', '0.2', '0.05', '0.04', '0.03', '0.02', '0.01',
       '&lt;0.01'], dtype=object)

In [35]:
df_merge2['Number of deaths due to HIV/AIDS'].unique()

array(['690000', '730000', '760000', '800000', '830000', '870000',
       '930000', '1000000', '1100000', '1200000', '1400000', '1500000',
       '1600000', '1700000', '1300000', '940000', '700000', '580000',
       '480000', '380000', '300000', '440000', '460000', '510000',
       '530000', '570000', '610000', '680000', '740000', '810000',
       '890000', '910000', '710000', '520000', '430000', '350000',
       '280000', '220000', '52000', '50000', '51000', '54000', '56000',
       '57000', '59000', '63000', '65000', '67000', '69000', '75000',
       '78000', '79000', '80000', '73000', '71000', '98000', '110000',
       '99000', '89000', '66000', '120000', '130000', '140000', '150000',
       '170000', '190000', '200000', '230000', '210000', '100000',
       '83000', '64000', '46000', '32000', '20000', '12000', '5800',
       '39000', '38000', '37000', '34000', '33000', '36000', '31000',
       '28000', '25000', '22000', '18000', '17000', '19000', '26000',
       '24000', '16000', '1

**Correcting The values**

In [36]:
for i in df_merge2.columns:
    if i=='New HIV infections (per 1000 uninfected population) (Both Sexes)' or i=='New HIV infections (per 1000 uninfected population) (Female)':
        df_merge2[i]=df_merge2[i].str.replace('&lt;0.01','0.01')

In [37]:
df_merge2['Number of deaths due to HIV/AIDS'] = df_merge2['Number of deaths due to HIV/AIDS'].str.replace('&lt;500','500')

***checking now***

In [38]:
df_merge2['New HIV infections (per 1000 uninfected population) (Female)'].unique()

array(['0.21', '0.22', '0.23', '0.24', '0.26', '0.27', '0.28', '0.29',
       '0.31', '0.32', '0.34', '0.35', '0.36', '0.37', '0.39', '0.4',
       '0.41', '0.43', '0.45', '0.47', '0.48', '0.49', '0.5', '0.44',
       '1.11', '1.18', '1.29', '1.38', '1.48', '1.6', '1.71', '1.85',
       '1.99', '2.14', '2.26', '2.39', '2.53', '2.66', '2.79', '2.92',
       '3.07', '3.23', '3.4', '3.56', '3.71', '3.84', '3.95', '4.02',
       '4.07', '4.08', '3.99', '3.83', '3.55', '3.19', '0.09', '0.1',
       '0.11', '0.12', '0.13', '0.14', '0.15', '0.06', '0.07', '0.08',
       '0.16', '0.19', '0.2', '0.05', '0.04', '0.03', '0.02', '0.01'],
      dtype=object)

In [39]:
df_merge2['New HIV infections (per 1000 uninfected population) (Both Sexes)'].unique()

array(['0.22', '0.23', '0.24', '0.25', '0.26', '0.27', '0.28', '0.29',
       '0.3', '0.32', '0.33', '0.34', '0.35', '0.36', '0.37', '0.38',
       '0.4', '0.42', '0.43', '0.45', '0.47', '0.48', '0.49', '0.41',
       '0.94', '1', '1.09', '1.17', '1.26', '1.36', '1.46', '1.57', '1.7',
       '1.83', '1.94', '2.06', '2.18', '2.3', '2.41', '2.53', '2.66',
       '2.8', '2.94', '3.09', '3.22', '3.34', '3.44', '3.51', '3.57',
       '3.59', '3.54', '3.42', '3.21', '2.92', '0.17', '0.18', '0.19',
       '0.2', '0.21', '0.08', '0.09', '0.1', '0.11', '0.12', '0.13',
       '0.14', '0.15', '0.16', '0.07', '0.06', '0.05', '0.04', '0.03',
       '0.02', '0.01'], dtype=object)

In [40]:
df_merge2['Number of deaths due to HIV/AIDS'].unique()

array(['690000', '730000', '760000', '800000', '830000', '870000',
       '930000', '1000000', '1100000', '1200000', '1400000', '1500000',
       '1600000', '1700000', '1300000', '940000', '700000', '580000',
       '480000', '380000', '300000', '440000', '460000', '510000',
       '530000', '570000', '610000', '680000', '740000', '810000',
       '890000', '910000', '710000', '520000', '430000', '350000',
       '280000', '220000', '52000', '50000', '51000', '54000', '56000',
       '57000', '59000', '63000', '65000', '67000', '69000', '75000',
       '78000', '79000', '80000', '73000', '71000', '98000', '110000',
       '99000', '89000', '66000', '120000', '130000', '140000', '150000',
       '170000', '190000', '200000', '230000', '210000', '100000',
       '83000', '64000', '46000', '32000', '20000', '12000', '5800',
       '39000', '38000', '37000', '34000', '33000', '36000', '31000',
       '28000', '25000', '22000', '18000', '17000', '19000', '26000',
       '24000', '16000', '1

In [41]:
df_merge2.dtypes

WHO region                                                          object
Year                                                                object
Number of new HIV infections                                        object
New HIV infections (per 1000 uninfected population) (Both Sexes)    object
New HIV infections (per 1000 uninfected population) (Male)          object
New HIV infections (per 1000 uninfected population) (Female)        object
Number of deaths due to HIV/AIDS                                    object
PEOPLE HAVING HIV                                                   object
dtype: object

**Correcting The Datatypes**

In [42]:
df_merge2['Number of new HIV infections'] = df_merge2['Number of new HIV infections'].astype(str).astype(int)

df_merge2['New HIV infections (per 1000 uninfected population) (Both Sexes)'] = df_merge2['New HIV infections (per 1000 uninfected population) (Both Sexes)'].astype(str).astype(float)

df_merge2['New HIV infections (per 1000 uninfected population) (Male)'] = df_merge2['New HIV infections (per 1000 uninfected population) (Male)'].astype(str).astype(float)

df_merge2['New HIV infections (per 1000 uninfected population) (Female)'] = df_merge2['New HIV infections (per 1000 uninfected population) (Female)'].astype(str).astype(float)

df_merge2['Number of deaths due to HIV/AIDS'] = df_merge2['Number of deaths due to HIV/AIDS'].astype(str).astype(int)

df_merge2['PEOPLE HAVING HIV'] = df_merge2['PEOPLE HAVING HIV'].astype(str).astype(int)


In [43]:
df_merge2.dtypes

WHO region                                                           object
Year                                                                 object
Number of new HIV infections                                          int32
New HIV infections (per 1000 uninfected population) (Both Sexes)    float64
New HIV infections (per 1000 uninfected population) (Male)          float64
New HIV infections (per 1000 uninfected population) (Female)        float64
Number of deaths due to HIV/AIDS                                      int32
PEOPLE HAVING HIV                                                     int32
dtype: object

# Exporting The Final Sheet For Visualization In Tableau

In [44]:
df_merge2.to_excel('HIV_FINAL.xlsx')